### ISSUE TITLE
The claim is that the first neighbor is more frequently missing than other neighbors when search for already indexed elements.

### ISSUE THREAD

**Issue by stepanv1** [nmslib/nmslib#337]
Code run:

import nmslib
import numpy as np

data = np.random.randn(100000, 100).astype(np.float32)

index = nmslib.init(method='hnsw', space='l2')
index.addDataPointBatch(data)
index.createIndex({'post': 2}, print_progress=True)

neighbours = index.knnQueryBatch(data, k=100, num_threads=4)

Produces wrong neighbour indices, e.g.:
In[18]: neighbours[26]
Out[18]: 
(array([ 1425, 91539, 89951, 28026, 24791, 92831, 22923, 38774, 50008,
        64587, 11995, 27318, 58637, 25437, 75790, 19562, 56337, 17156,
        95528,  1006, 65472, 38453, 79941,  7821, 30155, 12159, 22305,
        44764, 48253,  3574, 62078, 77389, 75657,  7773, 19723, 57821,
        46659, 39697, 83586, 93484, 94313, 58600, 37974, 73261, 99305,
        27483, 90557,  1701, 18909, 18693, 58646, 60283, 81228, 52749,
         3199, 22266, 17276, 79445, 42455, 41112, 33097, 33375, 65903,
        11406, 79207,  4517, 42406, 76789, 55040,  1975, 30976, 97678,
        16839,  3651, 99147, 62360, 91618, 63492, 97220, 45700, 62413,
        76402, 90970, 19880,  1476, 62582,  1911,  4781, 56462, 33622,
        37130, 10276, 88494, 98756, 14765, 29797,  3151, 30094, 65032,
        73954], dtype=int32),
 array([10.708295 , 10.740475 , 10.894506 , 10.967306 , 11.0539255,
        11.069767 , 11.128398 , 11.133954 , 11.212067 , 11.218391 ,
        11.234944 , 11.235495 , 11.262772 , 11.29688  , 11.297857 ,
        11.2993145, 11.309425 , 11.311992 , 11.343375 , 11.387876 ,
        11.405055 , 11.418722 , 11.420082 , 11.425428 , 11.435059 ,
        11.435092 , 11.442741 , 11.444837 , 11.454805 , 11.463407 ,
        11.4700575, 11.473954 , 11.494736 , 11.520248 , 11.526887 ,
        11.5283785, 11.535189 , 11.542302 , 11.562851 , 11.621191 ,
        11.640388 , 11.650836 , 11.652021 , 11.653956 , 11.66735  ,
        11.668255 , 11.6754055, 11.675482 , 11.677728 , 11.680329 ,
        11.693094 , 11.704804 , 11.709594 , 11.717174 , 11.725717 ,
        11.735891 , 11.740176 , 11.743173 , 11.7459135, 11.75041  ,
        11.751218 , 11.7527485, 11.75385  , 11.755196 , 11.761931 ,
        11.764097 , 11.77444  , 11.775189 , 11.7752495, 11.778778 ,
        11.782919 , 11.783841 , 11.791599 , 11.794125 , 11.795544 ,
        11.812305 , 11.820616 , 11.823801 , 11.8252735, 11.826242 ,
        11.82943  , 11.845283 , 11.84622  , 11.848603 , 11.876535 ,
        11.877209 , 11.880156 , 11.88081  , 11.881638 , 11.88944  ,
        11.88949  , 11.893159 , 11.89662  , 11.898159 , 11.902648 ,
        11.903265 , 11.911263 , 11.91164  , 11.921512 , 11.933577 ],
       dtype=float32))

Instead of 1425 first index should be 0. Bug appears in different points after multiple runs.
With decreasing of the number of dimensions number of errors decreases.

**Comment by searchivarius** [comment]
@stepanv1 why should the index be 0?

**Comment by stepanv1** [comment]
@searchivarius Sorry, typo. Should the index of the element. In the case above --- 26

**Comment by searchivarius** [comment]
@stepanv1 is 26 a true neighbor? In any case, there's no guarantee that all neighbors are returned. It's an approximate search method.

**Comment by stepanv1** [comment]
@searchivarius Yes, 26 is a true neighbour. Point is that the first neighbour in the output is always the index of the element itself, since the closest element to the element with e.g., index 26, is the element with index 26. So no matter how bad is the approximation, especially at higher dimensions, I suspect this being a bug and not the algorithm's property. 
Normally, in my runs of this and other versions of this approximated k-nn search algorithm , errors are made at the computation of mostly far-away-neighbours, with k>20.

**Comment by searchivarius** [comment]
@stepanv1 I really doubt it is a bug. Random data is quite difficult already for d > 10 and it's 100 in your case. All data points are roughly at the same distance from each other. For high dimensional data nearest neighbor search even doesn't make sense: only the near duplicate search makes sense.

**Comment by stepanv1** [comment]
@yurymalkov @searchivarius 
Thank you for your advice, I tried that ( (e.g. index.createIndex({'post': 2,'efConstruction':'1000','M':64}, print_progress=True))) before and it works, reducing number of errors. But I am not talking about near duplicates. I mean exact identical case, when the first nearest neighbour of any vector should be the vector itself. If algorithm does not return as 1 n.n. for the vector number 26 the vector 26 itself, may be this is something worth of checking? Normal way this approximate algorithm deviates from exact, according to my observations,  is in the farthest neighbours, returning correct 1st nearest neigbours. I do not think that this is related to the randomness of data. Here are the two examples to demonstrate it:
     
Here is the example of code, giving about 300 errors in 1st nearest neigbor, random datat:

data = np.random.randn(100000, 100).astype(np.float32)
index = nmslib.init(method='hnsw', space='l2')
index.addDataPointBatch(data)
index.createIndex({'post': 2,'efConstruction':'1000','M':64}, print_progress=True)
neighbours = index.knnQueryBatch(data, k=10, num_threads=40)
err=[neighbours[i][0][0]!=i for i in  range(len(neighbours))]
sum(err)

Now, generate blobs to create same size of data in previous, but clustered, still about 300 errors in the first neigbour:

from sklearn.datasets.samples_generator import make_blobs
data, y = make_blobs(n_samples=100000, centers=20, n_features=100, random_state=0)
data = np.random.randn(100000, 100).astype(np.float32)
# initialize a new index, using a HNSW index on Cosine Similarity
index = nmslib.init(method='hnsw', space='l2')
index.addDataPointBatch(data)
index.createIndex({'post': 2,'efConstruction':'1000','M':64}, print_progress=True)

neighbours = index.knnQueryBatch(data, k=10, num_threads=40)
err=[neighbours[i][0][0]!=i for i in  range(len(neighbours))]
sum(err)

**Comment by searchivarius** [comment]
Hi @stepanv1 thank you for the feedback.  Ok, did you actually check the recall for 1st neighbor vs nth neighbor for n > 1. I just ran a test with the experiment utility (but not with the Python API). I used data set points as queries: recall for the first element is higher as expected.

**Comment by stepanv1** [comment]
@searchivarius, I did not check the recall numbers, sorry. But I noticed that some of the output rows are simply shifted to the one position left, causing the disappearence of 1st nn, making 2nd n.n the 1st one etc. I hope that helps.

**Comment by searchivarius** [comment]
@stepanv1 which value of the parameter ef did you use? for large ef the recall is close to 100% overall, so most elements are present. so, without a proper measurement, this could have easily been a perception bias. most elements are present, but your mind could have focused on the few cases where the first element was missing. This is not unreasonable given that probabilities of missing the 1st and the 10th neighbor are pretty close.

**Comment by stepanv1** [comment]
@searchivarius, no it is not a perception bias. Rows were simply shifted  as a whole, I checked that. My workaround was  shifting the row back to the right, using hnswlib:
import hnswlib
from scipy.ndimage.interpolation import shift
def find_neighbors(data, k_, metric='euclidean', cores=12):
   dim = np.shape(data)[1]
   num_elements = np.shape(data)[0]
   data_labels = np.arange(num_elements)
   p = hnswlib.Index(space='l2', dim=dim)
   p.init_index(max_elements=num_elements, ef_construction=5000, M=16)
   
   p.add_items(data, data_labels)
 
   p.set_ef(5000) 

   labels, distances = p.knn_query(data, k=k_+1)
   #correct results
   IDX = distances[:,0]!=0
   distances = np.array([shift(distances[i,:], 1, cval=0) if IDX[i] else distances[i,:] for i in data_labels ])
   labels = np.array([shift(labels[i, :], 1, cval=0) if IDX[i] else labels[i, :] for i in data_labels])
   return {'dist':distances[:,1:], 'idx':labels[:,1:]}

**Comment by stepanv1** [comment]
Yes, agree, still needs to be checked if its the same problem, don't have time right now.

**Comment by searchivarius** [comment]
Ok, I will check when I have time, although I really-really doubt this, especially without a reproducible example with the properly measured outcome.

**Comment by searchivarius** [comment]
@stepanv1 and @yurymalkov I have suddenly decided to be done with it, so, I did create a script to compute recall for various n-th neighbors: https://github.com/nmslib/nmslib/blob/ac5febb1a24d7a5efa797b178dcdbfb4b5462d69/python_bindings/notebooks/test_hnsw_recall.py

Here is what I get for random data. No surprises here, for the vary accurate search (``efS=1000``), we do find all the true nearest neighbors in this example:
```
kNN recall for neighbor 1 1.000000
kNN recall for neighbor 2 1.000000
kNN recall for neighbor 3 0.900000
kNN recall for neighbor 4 0.900000
kNN recall for neighbor 5 0.800000
kNN recall for neighbor 6 0.800000
kNN recall for neighbor 7 0.700000
kNN recall for neighbor 8 0.600000
kNN recall for neighbor 9 0.500000
kNN recall for neighbor 10 0.200000
```
For ``efS=100`` we sometimes miss the 1-st neighbor and the 2-d neighbor, but less frequently than other neighbors. @stepanv1 feel free to run this test with your data:
```
kNN recall for neighbor 1 0.900000
kNN recall for neighbor 2 0.900000
kNN recall for neighbor 3 0.900000
kNN recall for neighbor 4 0.800000
kNN recall for neighbor 5 0.800000
kNN recall for neighbor 6 0.800000
kNN recall for neighbor 7 0.500000
kNN recall for neighbor 8 0.500000
kNN recall for neighbor 9 0.100000
kNN recall for neighbor 10 0.000000
```
